src/circuits/noise.jl

```diff
@@ -1,91 +1,73 @@
-# Noise model gate definitions
-function gate(::GateName"1qubit-pauli_error"; pX::Number = 0.0, pY::Number = 0.0, pZ::Number = 0.0)
-  kraus = zeros(Complex{Float64},2,2,4)
-  kraus[:,:,1] = √(1-pX-pY-pZ) * gate("Id")
-  kraus[:,:,2] = √pX * gate("X")
-  kraus[:,:,3] = √pY * gate("Y")
-  kraus[:,:,4] = √pZ * gate("Z")
-  return kraus 
-end
 
-function gate(::GateName"2qubit-pauli_error"; probs = prepend!(zeros(15),1))
-  kraus = zeros(Complex{Float64},4,4,16)
-  basis = vec(reverse.(Iterators.product(fill(["Id","X","Y","Z"],2)...)|>collect))
+function gate(::GateName"pauli_channel", N::Int = 1; 
+              error_probabilities = prepend!(zeros(4^N-1),1),
+              pauli_ops = ["Id","X","Y","Z"]) 
+  length(error_probabilities) > (1 << 10) && error("Hilbert space too large")
+  error_probabilities ./= sum(error_probabilities)
+  kraus = zeros(Complex{Float64},1<<N,1<<N,4^N)
+  basis = vec(reverse.(Iterators.product(fill(pauli_ops,N)...)|>collect))
   for (k,ops) in enumerate(basis)
-    kraus[:,:,k] = √probs[k] * kron(gate(ops[1]), gate(ops[2]))
+    kraus[:,:,k] = √error_probabilities[k] * reduce(kron,gate.(ops)) 
   end
-  return kraus 
-end
-
-function gate(::GateName"pauli_error", N::Int = 1; kwargs...) 
-  N == 1 && return gate("1qubit-pauli_error"; kwargs...)
-  N == 2 && return gate("2qubit-pauli_error"; kwargs...)
+  return kraus
 end
 
+gate(::GateName"bit_flip", N::Int = 1; p::Number) = 
+  gate("pauli_channel", N; error_probabilities = prepend!(p/(4^N-1) * ones(4^N-1), 1-p), pauli_ops = ["Id","X"])
 
-gate(::GateName"pauli_channel"; kwargs...) = gate("pauli_error"; kwargs...)
-
-
-gate(::GateName"bit_flip"; p::Number) = 
-  gate("pauli_error"; pX = p)[:,:,1:2]
-  
-gate(::GateName"phase_flip"; p::Number) = 
-  gate("pauli_error"; pZ = p)[:,:,[1,4]]
-
-gate(::GateName"bit_phase_flip"; p::Number) = 
-  gate("pauli_error"; pY = p)[:,:,[1,3]]
-
-gate(::GateName"phase_bit_flip"; kwargs...) = 
-  gate("bit_phase_flip"; kwargs...)
+gate(::GateName"phase_flip", N::Int = 1; p::Number) = 
+  gate("pauli_channel", N; error_probabilities = prepend!(p/(4^N-1) * ones(4^N-1), 1-p), pauli_ops = ["Id","Z"])
 
-function gate(::GateName"AD"; γ::Number)
+function gate(::GateName"AD", N::Int = 1; γ::Number)
+  K = zeros(Complex{Float64},1<<N,1<<N,2^N)
```

Comment:
Does this need to be complex? In general is `γ > 0` or `< 0`?

If it could be both, it might be preferable to just make it type unstable, i.e.:
```julia
ElType = γ >= 0 ? Float64 : ComplexF64
K = zeros(ElType,1<<N,1<<N,2^N)
[...]
```